src/arch_mips.py

```diff
@@ -606,35 +608,79 @@ def normalize_instruction(cls, instr: AsmInstruction) -> AsmInstruction:
     def parse(
         cls, mnemonic: str, args: List[Argument], meta: InstructionMeta
     ) -> Instruction:
+        inputs: List[Access] = []
+        outputs: List[Access] = []
+        clobbers: List[Access] = []
         jump_target: Optional[Union[JumpTarget, Register]] = None
         function_target: Optional[Union[AsmGlobalSymbol, Register]] = None
         has_delay_slot = False
         is_branch_likely = False
         is_conditional = False
         is_return = False
 
+        memory_sizes = {
+            "b": 1,
+            "h": 2,
+            "w": 4,
+            "d": 8,
+        }
+        size = memory_sizes.get(mnemonic[1:2])
+
+        def make_memory_access(arg: Argument) -> Access:
+            assert size is not None
+            assert not isinstance(arg, Register)
+            if isinstance(arg, AsmAddressMode):
+                return MemoryAccess(
+                    base_reg=arg.rhs,
+                    offset=arg.lhs,
+                    size=size,
+                )
+            return MemoryAccess(
+                base_reg=Register("zero"),
+                offset=arg,
+                size=size,
+            )
+
         if mnemonic == "jr" and args[0] == Register("ra"):
             # Return
+            assert len(args) == 1
+            inputs = [Register("ra")]
             is_return = True
             has_delay_slot = True
         elif mnemonic == "jr":
             # Jump table (switch)
-            assert isinstance(args[0], Register)
+            assert len(args) == 1 and isinstance(args[0], Register)
+            inputs.append(args[0])
             jump_target = args[0]
             is_conditional = True
             has_delay_slot = True
         elif mnemonic == "jal":
             # Function call to label
-            assert isinstance(args[0], AsmGlobalSymbol)
+            assert len(args) == 1 and isinstance(args[0], AsmGlobalSymbol)
+            inputs = list(cls.argument_regs)
+            outputs = list(cls.all_return_regs)
+            clobbers = list(cls.temp_regs)
+            clobbers.append(MemoryAccess.arbitrary())
             function_target = args[0]
             has_delay_slot = True
         elif mnemonic == "jalr":
             # Function call to pointer
-            assert isinstance(args[0], Register)
-            function_target = args[0]
+            assert (
+                len(args) == 2
+                and args[0] == Register("ra")
+                and isinstance(args[1], Register)
+            )
+            inputs = list(cls.argument_regs)
+            inputs.append(Register("ra"))
```

Comment:
we don't read from ra, do we?

Reply:
translate.py doesn't track ra, but I thought it might be more consistent to still list it as an input here? Similarly, it'd be marked as an input in `lw    $ra, 0x14($sp)` or an output in `lw $ra, 0x14($sp)` -- despite also being ignored there too.

Reply:
I generally agree, but jalr writes to ra, it doesn't read it. And that write is then immediately clobbered by the function call.

Reply:
My bad! I was thinking about `jr $ra`! I'll remove it